python-package/securexgboost/remote_server.py

```diff
@@ -416,16 +430,21 @@ def _synchronize(self, func, params):
         username = params.username
 
         self.condition.acquire() 
-        self.command.submit(func, params, username)
-        if self.command.is_ready():
-            self.command.invoke()
-            ret = self.command.result(username)
+        try:
+            self.command.submit(func, params, username)
+            if self.command.is_ready():
+                self.command.invoke()
+                ret = self.command.result(username)
+                self.condition.notifyAll()
+            else:
+                self.condition.wait()
+                ret = self.command.result(username)
+            self.condition.release()
+            return ret
+        except:
             self.condition.notifyAll()
-        else:
-            self.condition.wait()
-            ret = self.command.result(username)
-        self.condition.release()
-        return ret
+            self.condition.release()
+            raise Exception(self.command._error)
```

Comment:
Will this line raise an exception for all threads or only for this thread? i.e. if a thread is waiting as on line 440 when another thread raises an exception, will the waiting thread also raise an exception or be notified that an exception should be thrown, perhaps as part of the result on line 441? 

Or should `self.error` also be set here to communicate that an error was thrown to all other waiting threads?

Reply:
For all threads.
A thread that is waiting on line 440 will be woken up by lines 445-446 when some thread T raises an exception. When it tries to retrieve the result on line 441, `self.command.result` will throw an exception (because the previous thread will have already set `self._error`)

Although, it seems like `self._error` never gets reset again... Let me fix this before you merge.

Reply:
Fixed.